File: backend/src/routes/sync.ts
import type { FastifyInstance } from "fastify";
import { query } from "../db";
import { NotFoundError } from "../errors";
import { getCurrentMembership, getCurrentUser, requireCurrentUser } from "../middleware/auth";
import {
  SQL_SELECT_SYNC_JOB_BY_ID,
  SQL_SELECT_SYNC_JOB_STREAM,
  buildSyncJobsListSql,
} from "../sql/sync-jobs";
import { getOrgBySlug } from "../sql/orgs";
import { mapSyncJob } from "../utils/serialization";

export default async function syncRoutes(api: FastifyInstance) {
  api.get("/sync/jobs", { preHandler: requireCurrentUser }, async (request) => {
    const currentUser = getCurrentUser(request);
    const { org, connector_id: connectorId, limit = "50", offset = "0" } = request.query as Record<string, string>;
    const organization = await getOrgBySlug(org);
    await getCurrentMembership(org, currentUser.id);
    const values: unknown[] = [organization.id];
    if (connectorId) {
      values.push(connectorId);
    }
    values.push(Number(limit), Number(offset));
    const limitPos = connectorId ? 3 : 2;
    const offsetPos = connectorId ? 4 : 3;
    const result = await query(
      buildSyncJobsListSql(!!connectorId, limitPos, offsetPos),
      values
    );
    return result.rows.map(mapSyncJob);
  });

  api.get("/sync/jobs/:jobId", { preHandler: requireCurrentUser }, async (request) => {
    const currentUser = getCurrentUser(request);
    const { jobId } = request.params as { jobId: string };
    const { org } = request.query as { org: string };
    const organization = await getOrgBySlug(org);
    await getCurrentMembership(org, currentUser.id);
    const result = await query(SQL_SELECT_SYNC_JOB_BY_ID, [jobId, organization.id]);
    if (!result.rowCount) throw new NotFoundError("Sync job not found");
    return mapSyncJob(result.rows[0]);
  });

  api.get("/sync/jobs/:jobId/stream", { preHandler: requireCurrentUser }, async (request, reply) => {
    const currentUser = getCurrentUser(request);
    const { jobId } = request.params as { jobId: string };
    const { org } = request.query as { org: string };
    const organization = await getOrgBySlug(org);
    await getCurrentMembership(org, currentUser.id);

    reply.raw.writeHead(200, {
      "Content-Type": "text/event-stream",
      "Cache-Control": "no-cache",
      Connection: "keep-alive",
      "X-Accel-Buffering": "no",
    });

    const terminalStates = new Set(["done", "failed"]);
    let iterations = 0;

    const interval = setInterval(async () => {
      iterations += 1;
      const result = await query(SQL_SELECT_SYNC_JOB_STREAM, [jobId, organization.id]);
      const job = result.rows[0];
      if (!job) {
        reply.raw.write(`event: error\ndata: ${JSON.stringify({ detail: "Job not found" })}\n\n`);
        clearInterval(interval);
        reply.raw.end();
        return;
      }
      reply.raw.write(`data: ${JSON.stringify(job)}\n\n`);
      if (terminalStates.has(String(job.status)) || iterations >= 3600) {
        clearInterval(interval);
        reply.raw.end();
      }
    }, 1000);

    request.raw.on("close", () => clearInterval(interval));
    return reply;
  });
}

File: backend/app/routes/sync.py
from __future__ import annotations

import asyncio
import json
from typing import Any, AsyncGenerator, Optional

from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse
from sqlalchemy import select

import app.sql.sync_jobs as sql_jobs
from app.db.pool import get_pool
from app.db.tables import organizations, sync_jobs as sync_jobs_t
from app.errors import NotFoundError
from app.middleware.auth import get_current_membership, get_current_user

router = APIRouter()


def _map_sync_job(row: Any) -> dict:
    return {
        "id": str(row["id"]),
        "connector_id": str(row["connector_id"]),
        "org_id": str(row["org_id"]),
        "triggered_by": row.get("triggered_by"),
        "status": row["status"],
        "started_at": row["started_at"].isoformat() if row.get("started_at") else None,
        "finished_at": row["finished_at"].isoformat() if row.get("finished_at") else None,
        "docs_enqueued": row.get("docs_enqueued"),
        "docs_processed": row.get("docs_processed"),
        "docs_indexed": row.get("docs_indexed"),
        "docs_skipped": row.get("docs_skipped"),
        "docs_errored": row.get("docs_errored"),
        "error_message": row.get("error_message"),
        "meta": row.get("meta"),
        "created_at": row["created_at"].isoformat() if row.get("created_at") else None,
    }


async def _get_org_id_for_user(slug: str, user_id: str) -> Any:
    membership = await get_current_membership(slug, user_id)
    return membership["org_id"]


@router.get("/sync/jobs")
async def sync_jobs_list(
    org: Optional[str] = Query(default=None),
    connector_id: Optional[str] = Query(default=None),
    limit: int = Query(default=20, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    current_user: dict = Depends(get_current_user),
) -> list:
    if not org:
        return []

    membership = await get_current_membership(org, str(current_user["id"]))
    org_id = membership["org_id"]

    pool = await get_pool()

    if connector_id:
        from app.db.tables import sync_jobs as sj

        stmt = (
            select(
                sj.c.id,
                sj.c.connector_id,
                sj.c.org_id,
                sj.c.triggered_by,
                sj.c.status,
                sj.c.started_at,
                sj.c.finished_at,
                sj.c.docs_enqueued,
                sj.c.docs_processed,
                sj.c.docs_indexed,
                sj.c.docs_skipped,
                sj.c.docs_errored,
                sj.c.error_message,
                sj.c.meta,
                sj.c.created_at,
            )
            .where(sj.c.org_id == org_id, sj.c.connector_id == connector_id)
            .order_by(sj.c.created_at.desc())
            .limit(limit)
            .offset(offset)
        )
        rows = await pool.fetch(stmt)
    else:
        rows = await pool.fetch(sql_jobs.select_sync_jobs_by_org(org_id, limit, offset))

    return [_map_sync_job(r) for r in rows]


@router.get("/sync/jobs/{job_id}")
async def sync_job_get(
    job_id: str,
    org_slug: Optional[str] = Query(default=None),
    current_user: dict = Depends(get_current_user),
) -> dict:
    pool = await get_pool()

    if org_slug:
        membership = await get_current_membership(org_slug, str(current_user["id"]))
        org_id = membership["org_id"]
    else:
        raw = await pool.fetchrow(
            select(sync_jobs_t.c.org_id).where(sync_jobs_t.c.id == job_id)
        )
        if raw is None:
            raise NotFoundError("Sync job not found")
        org_row = await pool.fetchrow(
            select(organizations.c.slug).where(organizations.c.id == raw["org_id"])
        )
        membership = await get_current_membership(org_row["slug"], str(current_user["id"]))
        org_id = membership["org_id"]

    row = await pool.fetchrow(sql_jobs.select_sync_job_by_id(job_id, org_id))
    if row is None:
        raise NotFoundError("Sync job not found")
    return _map_sync_job(row)


@router.get("/sync/jobs/{job_id}/stream")
async def sync_job_stream(
    job_id: str,
    org_slug: Optional[str] = Query(default=None),
    current_user: dict = Depends(get_current_user),
) -> StreamingResponse:
    pool = await get_pool()

    if org_slug:
        membership = await get_current_membership(org_slug, str(current_user["id"]))
        org_id = membership["org_id"]
    else:
        raw = await pool.fetchrow(
            select(sync_jobs_t.c.org_id).where(sync_jobs_t.c.id == job_id)
        )
        if raw is None:
            raise NotFoundError("Sync job not found")
        org_row = await pool.fetchrow(
            select(organizations.c.slug).where(organizations.c.id == raw["org_id"])
        )
        membership = await get_current_membership(org_row["slug"], str(current_user["id"]))
        org_id = membership["org_id"]

    async def event_generator() -> AsyncGenerator[str, None]:
        terminal = {"done", "failed", "cancelled"}
        while True:
            row = await pool.fetchrow(sql_jobs.select_sync_job_stream(job_id, org_id))
            if row is None:
                yield f"data: {json.dumps({'error': 'job not found'})}\n\n"
                break
            data = {
                "id": str(row["id"]),
                "status": row["status"],
                "docs_enqueued": row.get("docs_enqueued"),
                "docs_processed": row.get("docs_processed"),
                "docs_indexed": row.get("docs_indexed"),
                "docs_skipped": row.get("docs_skipped"),
                "docs_errored": row.get("docs_errored"),
                "error_message": row.get("error_message"),
                "started_at": row["started_at"].isoformat() if row.get("started_at") else None,
                "finished_at": row["finished_at"].isoformat() if row.get("finished_at") else None,
            }
            yield f"data: {json.dumps(data)}\n\n"
            if row["status"] in terminal:
                break
            await asyncio.sleep(1)

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


@router.post("/sync/jobs/{job_id}/cancel")
async def sync_job_cancel(
    job_id: str,
    org: Optional[str] = Query(default=None),
    current_user: dict = Depends(get_current_user),
) -> dict:
    pool = await get_pool()

    if org:
        membership = await get_current_membership(org, str(current_user["id"]))
        org_id = membership["org_id"]
    else:
        raw = await pool.fetchrow(
            select(sync_jobs_t.c.org_id).where(sync_jobs_t.c.id == job_id)
        )
        if raw is None:
            raise NotFoundError("Sync job not found")
        org_row = await pool.fetchrow(
            select(organizations.c.slug).where(organizations.c.id == raw["org_id"])
        )
        membership = await get_current_membership(org_row["slug"], str(current_user["id"]))
        org_id = membership["org_id"]

    row = await pool.fetchrow(sql_jobs.cancel_sync_job(job_id, org_id))
    if row is None:
        raise NotFoundError("Sync job not found or already finished")

    # Revoke associated Celery tasks so they stop if still queued/running
    from app.workers.celery_app import celery_app

    celery_app.control.revoke(f"sync-enumerate-{job_id}", terminate=True, signal="SIGTERM")
    celery_app.control.revoke(f"sync-finalize-{job_id}", terminate=True, signal="SIGTERM")

    return {"id": job_id, "status": "cancelled"}
